src/H5FDmpio.c

```diff
@@ -1662,6 +1678,1042 @@ H5FD__mpio_write(H5FD_t *_file, H5FD_mem_t type, hid_t H5_ATTR_UNUSED dxpl_id, h
+#if MPI_VERSION >= 3
+            if (MPI_SUCCESS != (mpi_code = MPI_Get_elements_x(&mpi_stat, buf_type, &bytes_read)))
+#else
+            if (MPI_SUCCESS != (mpi_code = MPI_Get_elements(&mpi_stat, MPI_BYTE, &bytes_read)))
+#endif
+                HMPI_GOTO_ERROR(FAIL, "MPI_Get_elements failed", mpi_code)
+        } /* end if */
+
+        /* If the rank0-bcast feature was used, broadcast the # of bytes read to
+         * other ranks, which didn't perform any I/O.
+         */
+        /* NOTE: This could be optimized further to be combined with the broadcast
+         *          of the data.  (QAK - 2019/1/2)
+         *       Or have rank 0 clear the unread parts of the buffer prior to
+         *          the bcast.  (NAF - 2021/9/15)
+         */
+        if (rank0_bcast)
+#if MPI_VERSION >= 3
+            if (MPI_SUCCESS != MPI_Bcast(&bytes_read, 1, MPI_COUNT, 0, file->comm))
+#else
+            if (MPI_SUCCESS != MPI_Bcast(&bytes_read, 1, MPI_INT, 0, file->comm))
+#endif
+                HMPI_GOTO_ERROR(FAIL, "MPI_Bcast failed", 0)
+
+                /* Get the type's size */
+#if MPI_VERSION >= 3
+        if (MPI_SUCCESS != (mpi_code = MPI_Type_size_x(buf_type, &type_size)))
+#else
+        if (MPI_SUCCESS != (mpi_code = MPI_Type_size(buf_type, &type_size)))
+#endif
+            HMPI_GOTO_ERROR(FAIL, "MPI_Type_size failed", mpi_code)
+
+        /* Compute the actual number of bytes requested */
+        io_size = type_size * size_i;
+
+        /* Check for read failure */
+        if (bytes_read < 0 || bytes_read > io_size)
+            HGOTO_ERROR(H5E_IO, H5E_READERROR, FAIL, "file read failed")
+
+        /* Check for incomplete read */
+        n = io_size - bytes_read;
+        if (n > 0) {
+            i = (int)count - 1;
+
+            /* Iterate over sorted array in reverse, filling in zeroes to
+             * sections of the buffers that were not read to */
+            do {
+                HDassert(i >= 0);
+
+#if MPI_VERSION >= 3
+                io_size    = MIN(n, (MPI_Count)s_sizes[i]);
+                bytes_read = (MPI_Count)s_sizes[i] - io_size;
+#else
+                io_size    = MIN(n, (int)s_sizes[i]);
+                bytes_read = (int)s_sizes[i] - io_size;
+#endif
+                HDassert(bytes_read >= 0);
+
+                HDmemset((char *)s_bufs[i] + bytes_read, 0, (size_t)io_size);
+
+                n -= io_size;
+                i--;
+            } while (n > 0);
+        }
+    }
+    else if (count > 0) {
+        haddr_t max_addr   = HADDR_MAX;
+        hbool_t fixed_size = FALSE;
+        size_t  size;
+
+        /* The read is part of an independent operation. As a result,
+         * we can't use MPI_File_set_view() (since it it a collective operation),
+         * and thus there is no point in setting up an MPI derived type, as
+         * (to the best of my knowledge) MPI I/O doesn't have support for
+         * non-contiguous I/O in independent mode.
```

Comment:
This is incorrect - MPI_File_read_at can take any MPI datatype, including non-contiguous ones.

Reply:
I don't know if I have time for this now though. I can change the comment to reflect the situation.